flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CharVarCharTrimPadCastRule.java

```diff
@@ -103,27 +107,32 @@ protected String generateCodeBlockInternal(
 
             final CastRuleUtils.CodeWriter writer = new CastRuleUtils.CodeWriter();
             if (context.legacyBehaviour()
-                    || !(couldTrim(length) || couldPad(targetLogicalType, length))) {
+                    || ((!couldTrim(targetLength)
+                                    // Assume input length is respected by the source
+                                    || (inputLength != null && inputLength <= targetLength))
```

Comment:
this condition is already quite complex. maybe introduce another method that takes input from above and target length?